vllm/model_executor/layers/rotary_embedding/common.py

```diff
@@ -186,3 +118,137 @@ def _flashinfer_rotary_embedding_fake(
+    """
+    ro_dim = cos.shape[-1] * 2
+    assert ro_dim <= x.shape[-1]
+    cos = repeat(
+        cos, "... d -> ... 1 (2 d)" if not interleaved else "... d -> ... 1 (d 2)"
+    )
+    sin = repeat(
+        sin, "... d -> ... 1 (2 d)" if not interleaved else "... d -> ... 1 (d 2)"
+    )
+    return torch.cat(
+        [
+            x[..., :ro_dim] * cos + _rotate_half(x[..., :ro_dim], interleaved) * sin,
+            x[..., ro_dim:],
+        ],
+        dim=-1,
+    )
+
+
+def _apply_rotary_emb_torch(
+    x: torch.Tensor,
+    cos: torch.Tensor,
+    sin: torch.Tensor,
+    is_neox_style: bool,
+) -> torch.Tensor:
+    cos = cos.unsqueeze(-2).to(x.dtype)
+    sin = sin.unsqueeze(-2).to(x.dtype)
+    if is_neox_style:
+        x1, x2 = torch.chunk(x, 2, dim=-1)
+    else:
+        x1 = x[..., ::2]
+        x2 = x[..., 1::2]
+    o1 = x1 * cos - x2 * sin
+    o2 = x2 * cos + x1 * sin
+    if is_neox_style:
+        return torch.cat((o1, o2), dim=-1)
+    else:
+        return torch.stack((o1, o2), dim=-1).flatten(-2)
+
+
+@CustomOp.register("apply_rotary_emb")
+class ApplyRotaryEmb(CustomOp):
+    def __init__(
+        self,
+        is_neox_style: bool = False,
+        is_unsqueeze: bool = False,
+        default: Callable[..., torch.Tensor] = _apply_rotary_emb_torch,
+    ) -> None:
+        super().__init__()
+        self.default = default
+        self.is_neox_style = is_neox_style
+        self.is_unsqueeze = is_unsqueeze
+
+    @staticmethod
+    def forward_static(
+        x: torch.Tensor,
+        cos: torch.Tensor,
+        sin: torch.Tensor,
+        is_neox_style: bool = False,
+    ) -> torch.Tensor:
+        output = _apply_rotary_emb_torch(x, cos, sin, is_neox_style).type_as(x)
+        return output
+
+    def forward_native(
+        self,
+        x: torch.Tensor,
+        cos: torch.Tensor,
+        sin: torch.Tensor,
+    ) -> torch.Tensor:
+        output = _apply_rotary_emb_torch(x, cos, sin, self.is_neox_style).type_as(x)
+        return output
+
+    def forward_cuda(
+        self,
+        x: torch.Tensor,
+        cos: torch.Tensor,
+        sin: torch.Tensor,
+    ) -> torch.Tensor:
+        from vllm.vllm_flash_attn.layers.rotary import apply_rotary_emb
+
+        if self.is_unsqueeze:
+            output = apply_rotary_emb(
+                x.unsqueeze(0),
+                cos,
+                sin,
+                not self.is_neox_style,
+            ).squeeze(0)
+        else:
+            output = apply_rotary_emb(x, cos, sin).type_as(x)
+        return output
+
+    def forward_hip(
+        self,
+        x: torch.Tensor,
+        cos: torch.Tensor,
+        sin: torch.Tensor,
+    ) -> torch.Tensor:
+        apply_rotary_emb = None
+
+        # If torch compile is not enabled, use rotary embedding function from
+        # flash_attn package, otherwise use the naive pytorch embedding
+        # implementation is faster when torch compile is enabled.
+        if not torch.compiler.is_compiling():
```

Comment:
For follow-up: can we change this so the dispatch happens inside `__init__` (check `compilation_config.mode != CompilationMode.NONE`)?

Reply:
Actually, we should just always use FA in `forward_hip` as `forward_native` is used by default anyway when torch Inductor is used